src/clientiface.cpp

```diff
@@ -101,6 +101,17 @@ void RemoteClient::GetNextBlocks (
 	m_nothing_to_send_pause_timer -= dtime;
 	m_map_send_completion_timer += dtime;
 
+	if (m_map_send_completion_timer > g_settings->getFloat("server_unload_unused_data_timeout") * 0.8f) {
+		warningstream << "Server: Player " << m_name << ", RemoteClient " << peer_id
+				<< ": full map send is taking too long "
+				<< m_map_send_completion_timer
+				<< "s, restarting to avoid visible blocks being unloaded. \n"
+				<< "Consider increasing `server_unload_unused_data_timeout`"
+				<< std::endl;
```

Comment:
Without more careful tracking this will have a false positives because the map send starts anew when the camera or position changes.
I was careful with the wording suggestion in #13259 for this reason. It's not that map send _took_ so long, rather it _completed_ after this time since progress is not linear (okay maybe this is too subtle).

This log message is also too easily blaming the server settings for this, in theory a client that is slow to acknowledge block sends can also trigger this.

Reply:
Resetting the completion timer when camera/position change would fix this (and arguably would be the right thing to do anyway).

Reply:
> in theory a client that is slow to acknowledge block sends can also trigger this

True. But the remedy is the same: The server needs to cache the blocks for longer (or they'll be unloaded and cause inefficient behavior on both server and client)